src/reducers/console.js

```diff
@@ -33,22 +35,65 @@ export default function console(stateIn, {type, payload, meta}) {
           'evaluatedByCompiledProjectKey',
           payload.compiledProjectKey,
         ),
-      );
+      ));
     case 'EVALUATE_CONSOLE_ENTRY':
       return payload.trim(' ') === '' ? state : state.set(
-        meta.key,
-        new ConsoleEntry({expression: payload}),
-      );
+        'history', state.history.set(
+          meta.key,
+          new ConsoleEntry({expression: payload}),
+        ));
     case 'CLEAR_CONSOLE_ENTRIES':
       return initialState;
+    case 'CONSOLE_CHANGE':
+      return state.set('currentInputValue', payload.value);
     case 'CONSOLE_LOG_PRODUCED':
-      return state.set(
+      return state.set('history', state.history.set(
         meta.key,
         new ConsoleEntry({
           value: payload.value,
           evaluatedByCompiledProjectKey: payload.compiledProjectKey,
         }),
+      ));
+    case 'NAVIGATE_CONSOLE_HISTORY': {
+      const relevantHistory = state.history.toList().filter(
+        entry => entry.expression !== null,
       );
+
+      if (payload.direction === 'UP') {
+        if (relevantHistory.size === 0 ||
+            state.historyEntryIndex === relevantHistory.size) {
+          return state;
+        }
+        const firstHistoryNavigation = state.historyEntryIndex === null;
+        const newHistoryEntryIndex = firstHistoryNavigation ?
+          1 : state.historyEntryIndex + 1;
+        const historyIndex = relevantHistory.size - newHistoryEntryIndex;
+        const {expression} = relevantHistory.get(historyIndex);
+        const updatedState = state.
+          set('historyEntryIndex', newHistoryEntryIndex).
+          set('currentInputValue', expression);
+
+        return firstHistoryNavigation ?
+          updatedState.set('nextConsoleEntry', state.currentInputValue) :
+          updatedState;
+      }
+
+      if (state.historyEntryIndex === null) {
+        return state;
+      }
+      const newHistoryEntryIndex = state.historyEntryIndex - 1;
+      if (newHistoryEntryIndex === 0) {
+        return state.
+          set('historyEntryIndex', null).
+          set('nextConsoleEntry', null).
+          set('currentInputValue', state.nextConsoleEntry);
+      }
+      const historyIndex = relevantHistory.size - newHistoryEntryIndex;
+      const {expression} = relevantHistory.get(historyIndex);
+      return state.
+        set('historyEntryIndex', newHistoryEntryIndex).
+        set('currentInputValue', expression);
+    }
```

Comment:
So I think we can probably cut down the amount of code needed here if we don’t handle the up and down cases totally separately, something like:

* If the current history index is `null`, set it to 0.
* Then increment/decrement based on the value of `direction`
* Then check for each boundary case and handle appropriately
* Then handle the non-boundary case

Perhaps also some tweaks to how we represent the state overall might make this operation simpler, namely:

* Go back to representing the default history index as 0
* Keep the `nextConsoleEntry` up to date all the time (in response to the `CONSOLE_CHANGE` action, as long as the index is at 0)

Just throwing ideas out there—not 100% sure that they will simplify the code but maybe worth playing with!